cabal-install/Distribution/Client/CmdUpdate.hs

```diff
@@ -0,0 +1,179 @@
+    let repos       = filter isRepoRemote $ repoContextRepos repoCtxt
+        repoName    = remoteRepoName . repoRemote
+        parseArg :: String -> IO UpdateRequest
+        parseArg s = case simpleParse s of
+          Just r -> pure r
+          Nothing -> die' verbosity $ "'new-update' unable to parse repo: \"" ++ s ++ "\""
+    updateRepoRequests <- mapM parseArg extraArgs
+
+    unless (null updateRepoRequests) $ do
+      let remoteRepoNames = map repoName repos
+          unknownRepos = [r | (UpdateRequest r _) <- updateRepoRequests
+                            , not (r `elem` remoteRepoNames)]
+      unless (null unknownRepos) $
+        die' verbosity $ "'new-update' repo(s): \"" ++ intercalate "\", \"" unknownRepos
+                      ++ "\" can not be found in known remote repo(s): " ++ intercalate ", " remoteRepoNames
+
+    let reposToUpdate = case updateRepoRequests of
+          [] -> repos
+          updateRequests -> let repoNames = map _updateRequestRepoName updateRequests
+                            in filter (\r-> repoName r `elem` repoNames) repos
+
+    case reposToUpdate of
+      [] -> return ()
+      [remoteRepo] ->
+        notice verbosity $ "Downloading the latest package list from "
+                        ++ repoName remoteRepo
+      _ -> notice verbosity . unlines
+              $ "Downloading the latest package lists from: "
+              : map (("- " ++) . repoName) repos
+    jobCtrl <- newParallelJobControl (length reposToUpdate)
+    mapM_ (spawnJob jobCtrl . updateRepo verbosity defaultUpdateFlags repoCtxt) reposToUpdate
+    mapM_ (\_ -> collectJob jobCtrl) reposToUpdate
+
+  where
+    verbosity = fromFlagOrDefault normal (configVerbosity configFlags)
+    cliConfig = commandLineFlagsToProjectConfig
+                  globalFlags configFlags configExFlags
+                  installFlags haddockFlags
+
+updateRepo :: Verbosity -> UpdateFlags -> RepoContext -> Repo -> IO ()
+updateRepo verbosity updateFlags repoCtxt repo = do
+  transport <- repoContextGetTransport repoCtxt
+  case repo of
+    RepoLocal{..} -> return ()
+    RepoRemote{..} -> do
+      downloadResult <- downloadIndex transport verbosity repoRemote repoLocalDir
+      case downloadResult of
+        FileAlreadyInCache -> return ()
+        FileDownloaded indexPath -> do
+          writeFileAtomic (dropExtension indexPath) . maybeDecompress
+                                                  =<< BS.readFile indexPath
+          updateRepoIndexCache verbosity (RepoIndex repoCtxt repo)
+    RepoSecure{} -> repoContextWithSecureRepo repoCtxt repo $ \repoSecure -> do
+      let index = RepoIndex repoCtxt repo
+      -- NB: This may be a nullTimestamp if we've never updated before
+      current_ts <- currentIndexTimestamp (lessVerbose verbosity) repoCtxt repo
+      -- NB: always update the timestamp, even if we didn't actually
+      -- download anything
+      writeIndexTimestamp index (fromFlag (updateIndexState updateFlags))
```

Comment:
So that's the call where we'd need to propagate the `<index-state>` value to... what's the problem? :-)

Reply:
@angerman just to make this more explicit:

As you can see, `writeIndexTimestamp` writes the `01-index.timestamp` file which is stored alongside each `01-index.tar` in the package/repo cache:

```hs
-- | Write the 'IndexState' to the filesystem
writeIndexTimestamp :: Index -> IndexState -> IO () 
writeIndexTimestamp index st
  = writeFile (timestampFile index) (display st)
```

so all you need to do, is pass `writeIndexTimestamp` the optional `[,<index-state>]` argument associated with that specific repository.